### PR TITLE
Expression validation

### LINKED ISSUE #726: Validering ved hjelp av dynamikk
### Description

Vi ønsker en enklere måte å kunne sette opp valideringer på i skjema.
I dag må dette sette via validation eventet i C#, noe som gir mye og uoversiktlig kode, som igjen gir oss på sikt mye teknisk gjeld.
På samme måte som den nye dynamikken så ønsker vi en lignende løsning for valideringer.
Man må kunne sette opp regler som validere som true på lik linje som dynamikken, man bør kunne sjekke på et datamodellfelt eller en componentid.

- Valideringen skal ikke skje hvis felte har hidden=true.
- Man skal kunne benytte samme kontroller som ved dynamikk, dvs. "or", "equal" osv.
- Man må kunne ha muligheten til å linke inn et tekst id for å kunne ha en custom feilmeldingstekst.
- Man bør kunne styre om det er en hard kontroll eller en mjuk (error/warning)
- Man bør antagelig kunne si hvilke componentid som skal markers som feil, feks som ved summering av tall fra flere felter.
- Støtte warning og error
- Implementere alle metoder https://docs.altinn.studio/app/development/logic/expressions/

### Additional Information
Blocked by: https://github.com/Altinn/app-frontend-react/issues/1182

Includes cleanup in validation.ts ref. https://github.com/Altinn/app-frontend-react/issues/377

### PR DESCRIPTION
**Pull request by bjosttveit**
## Description

Validation using expressions. This feature requires app-lib v8, see: https://github.com/Altinn/app-lib-dotnet/pull/311.
- Fetches validation config `<dataType>.validation.json` from app backend.
- Runs expression validations along side other types of validations.
- Added argv expression to get positional arguments passed to an expression. (used to get the dataModel field the validation is for)

## Related Issue(s)

- closes #726 

## Verification/QA

- Manual functionality testing
  - [x] I have tested these changes manually
  - [ ] Creator of the original issue (or service owner) has been contacted for manual testing (or will be contacted when released in alpha)
  - [ ] No testing done/necessary
- Automated tests
  - [x] Unit test(s) have been added/updated
  - [x] Cypress E2E test(s) have been added/updated
  - [ ] No automatic tests are needed here (no functional changes/additions)
  - [ ] I want someone to help me make some tests
- UU/WCAG ([follow these guidelines](https://aksel.nav.no/god-praksis/artikler/utvikling?tema=universell-utforming) until we have our own)
  - [ ] I have tested with a screen reader/keyboard navigation/automated wcag validator
  - [x] No testing done/necessary (no DOM/visual changes)
  - [ ] I want someone to help me perform accessibility testing
- User documentation @ [altinn-studio-docs](https://github.com/Altinn/altinn-studio-docs)
  - [ ] Has been added/updated
  
  - [ ] No functionality has been changed/added, so no documentation is needed
  - [x] I will do that later/have created an issue
  
- Support in Altinn Studio
  - [ ] Issue(s) created for support in Studio
  
  - [ ] This change/feature does not require any changes to Altinn Studio
- Sprint board
  - [x] The original issue (or this PR itself) has been added to the Team Apps project and to the current sprint board
  - [ ] I don't have permissions to do that, please help me out
- Labels
  - [x] I have added a `kind/*` label to this PR for proper release notes grouping
  - [ ] I don't have permissions to add labels, please help me out

### CODE REVIEW

File: src/hooks/queries/useCustomValidationConfig.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  );
+  const instance: IInstance | null = useAppSelector((state) => state.instanceData.instance);
+  const layoutSets: ILayoutSets | null = useAppSelector((state) => state.formLayout.layoutsets);
+
+  const dataTypeId =
+    getCurrentDataTypeForApplication({
+      application: appMetadata,
+      instance,
+      layoutSets,
+    }) ?? '';
```
Comment (olemartinorg) [inline src/hooks/queries/useCustomValidationConfig.ts:31]: I've made a hook for this whole thing, `useCurrentDataModelName()` 🥳

File: src/utils/validation/expressionValidation.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    return null;
+  }
+
+  return resolvedDefinition as IExpressionValidationRefResolved;
+}
+
+/**
+ * Resolves a single expression validation definition.
+ */
+function resolveExpressionValidation(
```
Comment (olemartinorg) [inline src/utils/validation/expressionValidation.ts:57]: This function, and the one above it, seems to duplicate some things. Looks a bit strange, am I missing a reason the duplication is needed? 🤔

File: test/e2e/pageobjects/app-frontend.ts [hunk 3]
```diff
@@ -13,6 +13,9 @@ export class AppFrontend {
 
     /** @see https://dev.altinn.studio/repos/ttd/signing-test */
     signingTest: 'signing-test',
+
+    /** @see https://dev.altinn.studio/repos/ttd/expression-validation-test */
+    expressionValidationTest: 'expression-validation-test',
```
Comment (olemartinorg) [inline test/e2e/pageobjects/app-frontend.ts:18]: You should add this in the list of apps we test in the `README.md` as well. 🙌

**Review by olemartinorg — APPROVED** [review]
You're a mean, clean-coding machine, @bjosttveit! 🙌 🤩 🥳 Well done!!

File: schemas/json/layout/expression.schema.v1.json [hunk 4]
```diff
@@ -430,6 +431,16 @@
...
       ],
       "additionalItems": false
+    },
+    "func-argv": {
+      "title": "Positional argument function",
+      "description": "This function returns the value of the positional argument at the specified index passed to the expression (currently only available for validation conditions)",
+      "type": "array",
+      "items": [
+        { "const":  "argv" },
+        { "$ref": "#/definitions/string" }
```
[the remark this answers is not in the document]
Comment (bjosttveit) [reply schemas/json/layout/expression.schema.v1.json:441]: Oh yes of course 🤦

File: src/hooks/queries/useCustomValidationConfig.ts (hunk 1 above)
Comment (bjosttveit) [reply src/hooks/queries/useCustomValidationConfig.ts:31]: Oh nice!

File: test/e2e/pageobjects/app-frontend.ts (hunk 3 above)
Comment (bjosttveit) [reply test/e2e/pageobjects/app-frontend.ts:18]: 👍

File: src/utils/validation/expressionValidation.ts (hunk 2 above)
Comment (bjosttveit) [reply src/utils/validation/expressionValidation.ts:57]: Not entirely sure what you mean, but then again I have not looked at this function for weeks. I will take a look

Comment (olemartinorg) [reply src/utils/validation/expressionValidation.ts:57]: Both of them have things like:
```
 if (!('message' in ...)) {
    window.logWarn(`Custom validation:\nValidation for ${field} is missing a message.`);
    return null;
  }
```

**Comment by lassopicasso** [comment]
FYI: The following PR fixes the cy-test of grid; https://github.com/Altinn/app-frontend-react/pull/1551

File: src/utils/validation/expressionValidation.ts (hunk 2 above)
Comment (bjosttveit) [reply src/utils/validation/expressionValidation.ts:57]: They deal with the`validations` and `definitions` object (from the config) respectively, as they are treated slightly differently. But I have them both require a message and condition to be set at a minimum

**Review by olemartinorg — APPROVED** [review]

**Pull request merged**